src/serialization.jl

```diff
@@ -26,15 +70,11 @@ end
 Adds the object to the file and serializes it.
 """
 function Base.setindex!(jlso::JLSOFile, value, name::String)
-    io = IOBuffer()
-
-    if jlso.format === :bson
-        bson(io, Dict(name => value))
-    elseif jlso.format === :serialize
-        serialize(io, value)
-    else
-        error(LOGGER, ArgumentError("Unsupported format $(jlso.format)"))
-    end
+    buffer = IOBuffer()
+    compressing_buffer = compressor(jlso.compression).compress(buffer)
+    formatter(jlso.format).serialize!(compressing_buffer, value)
 
-    jlso.objects[name] = take!(io)
+    # need to close buffer so any compression can write end of body stuffs.
+    close(compressing_buffer)
+    jlso.objects[name] = buffer.data  # can't use take! as stream is now closed
```

Comment:
This seems gross, there must be a better way to get the end stuff in without closing the stream
look at 
https://github.com/bicycle1885/CodecZlib.jl/blob/a5ec201eba8447c5fd8c941d050ff5fa89ebd28f/src/compression.jl#L148
and
https://github.com/bicycle1885/CodecZlib.jl/blob/a5ec201eba8447c5fd8c941d050ff5fa89ebd28f/src/compression.jl#L148